content/200-orm/200-prisma-client/100-queries/090-raw-database-access/050-raw-queries.mdx

```diff
@@ -550,20 +524,210 @@ The database will thus provide a `String` representation of your data which Pris
+
+The following examples are all vulnerable ❌ to SQL Injection:
+
+```ts
+const inputString = `'Sarah' UNION SELECT id, title FROM "Post"`
+const result =
+  await prisma.$queryRaw`SELECT id, name FROM "User" WHERE name = ${Prisma.raw(
+    inputString
+  )}`
+console.log(result)
+```
+
+```ts
+const inputString = `'Sarah' UNION SELECT id, title FROM "Post"`
+const result = await prisma.$queryRaw(
+  Prisma.raw(`SELECT id, name FROM "User" WHERE name = ${inputString}`)
+)
+console.log(result)
+```
+
+```ts
+const inputString = `'Sarah' UNION SELECT id, title FROM "Post"`
+const query = Prisma.raw(
+  `SELECT id, name FROM "User" WHERE name = ${inputString}`
+)
+const result = await prisma.$queryRaw(query)
+console.log(result)
+```
+
+#### Safely using <inlinecode>$queryRaw</inlinecode> and <inlinecode>$executeRaw</inlinecode> in more complex scenarios
+
+##### Building raw queries separate to query execution
+
+If you want to build your raw queries elsewhere or separate to your parameters you will need to use one of the following methods.
+
+In this example, the `sql` helper method is used to build the query text by safely including the variable. It is safe ✅ from SQL Injection:
+
+```ts
+// inputString can be untrusted input
+const inputString = `'Sarah' UNION SELECT id, title FROM "Post"`
+
+// Safe if the text query below is completely trusted content
+const query = Prisma.sql`SELECT id, name FROM "User" WHERE name = ${inputString}`
+
+const result = await prisma.$queryRaw(query)
+console.log(result)
+```
+
+In these examples, the `sql` helper method is used to build the query text including a parameter marker for the input value. Each variable is represented by a marker symbol (`?` for mySQL, `$1`, `$2`, and so on for PostgreSQL).
+
+The following example uses a MySQL query and is safe ✅ from SQL Injection:
+
+```ts
+// Version for Typescript
+const query: any
+
+// Version for Javascript
+const query
+
+// Safe if the text query below is completely trusted content
+query = Prisma.sql`SELECT id, name FROM "User" WHERE name = ?`
+
+// inputString can be untrusted input
+const inputString = `'Sarah' UNION SELECT id, title FROM "Post"`
+query.values = [inputString]
+
+const result = await prisma.$queryRaw(query)
+console.log(result)
+```
+
+> **Note**: MySQL variables are represented by `?`
+
+The following example uses a PostgreSQL query and is safe ✅ from SQL Injection:
+
+```ts
+// Version for Typescript
+const query: any
+
+// Version for Javascript
+const query
+
+// Safe if the text query below is completely trusted content
+query = Prisma.sql`SELECT id, name FROM "User" WHERE name = $1`
+
+// inputString can be untrusted input
+const inputString = `'Sarah' UNION SELECT id, title FROM "Post"`
+query.values = [inputString]
+
+const result = await prisma.$queryRaw(query)
+console.log(result)
+```
+
+> **Note**: PostgreSQL variables are represented by `$1`, etc
+
+> **Note**: Subsequent examples will just use show PostgreSQL queries.
```

Comment:
Could we standardize on PostgreSQL throughout the page? Fine to mention `?` vs `$1` somewhere, but let’s keep examples to PostgreSQL.

Reply:
Sure I have made this change.